File: self_learn/stack.c
#include <stdio.h>

#define N 5
int stack[N];
int top = -1;

void push(int x)
{
    if (top == N - 1)
    {
        printf("Stack Overflow\n");
    }
    else
    {
        top++;
        stack[top] = x;
    }
}
void pop()
{
    if (top == -1)
    {
        printf("Stack Underflow\n");
    }
    else
    {
        printf("Removed element is %d\n", stack[top]);
        top--;
    }
}
void peek()
{
    if (top == -1)
    {
        printf("Stack Underflow\n");
    }
    else
    {
        printf("Top element is %d\n", stack[top]);
    }
}
void display()
{
    printf("\nElements from the top: \n");
    for (int i = top; i >= 0; i--)
    {
        printf("%d\n", stack[i]);
    }
}

int main()
{
    push(10);
    push(20);
    push(30);
    display();
    pop();
    peek();
    display();
    push(10);
    push(20);
    push(30);
    push(10);
    return 0;
}

File: self_learn/stack.py
stack = []
N = 5

def push(x):
    if len(stack) >= N:
        print("Stack Overlow")
    else:
        stack.append(x)
def pop():
    if not stack:
        print("Stack Underflow")
    else:
        print(f"Removed element is {stack[-1]}")
        del stack[-1]
        # or just write
        # stack.pop()
def peek():
    if not stack:
        print("Stack Underflow")
    else:
        print(f"Top element is {stack[-1]}")
def display():
    print("\nDisplaying from top: ")
    for i in reversed(stack):
        print(i)
        
push(10)
push(20)
push(30)
display()
pop()
peek()
display()
push(10)
push(20)
push(30)
push(10)

